datastore/batchwriter.go

```diff
@@ -56,9 +57,31 @@ func (bw *batchWriter) Write() error {
 		bwt := time.Now()
 		bw.stmtCount = 0
 		err := bw.dbClient.ExecuteBatch(bw.batch)
+		if err != nil {
+			// Retry with exponential backoff.
+			config.G.Log.System.LogWarn("Retrying MetricManager write...")
+			go bw.retryWrite()
+			logging.Statsd.Client.Inc("metricmgr.db.retry", 1, 1.0)
+		}
 		logging.Statsd.Client.TimingDuration("metricmgr.db.write", time.Since(bwt), 1.0)
-		return err
-	} else {
-		return nil
 	}
+	return nil
+}
+
+func (bw *batchWriter) retryWrite() error {
+	var i time.Duration
+	i = 0
+	var err error
+	for i < 5 {
+		err := bw.dbClient.ExecuteBatch(bw.batch)
```

Comment:
bw.batch is not stable; it will be overwritten by another goroutine.

You need to give this goroutine its own copy of everything that might change, in this case, the batch.